crates/holochain_websocket/src/lib.rs

```diff
@@ -688,13 +729,59 @@ impl WebsocketListener {
         Ok(Self {
             config,
             access_control,
-            listener,
+            listener: Box::new(listener),
+        })
+    }
+
+    /// Bind a new websocket listener on the same port using a v4 and a v6 socket.
+    pub async fn dual_bind(
+        config: Arc<WebsocketConfig>,
+        addr_v4: impl Into<SocketAddr>,
+        addr_v6: impl Into<SocketAddr>,
+    ) -> Result<Self> {
+        let access_control = Arc::new(config.allowed_origins.clone().ok_or_else(|| {
+            Error::other("WebsocketListener requires access control to be set in the config")
+        })?);
+
+        let mut addr_v4 = addr_v4.into();
+        let addr_v6 = addr_v6.into();
+
+        if !addr_v4.is_ipv4() {
+            return Err(Error::other("Expected an IPv4 address"));
+        }
+
+        if !addr_v6.is_ipv6() {
+            return Err(Error::other("Expected an IPv6 address"));
+        }
+
+        // Note that tokio binds to the stack matching the address type so we can re-use the port
+        // without needing to create the sockets ourselves to configure this.
+        let v6_listener = tokio::net::TcpListener::bind(addr_v6).await?;
+        addr_v4.set_port(v6_listener.local_addr()?.port());
+
+        let v4_listener = tokio::net::TcpListener::bind(addr_v4).await?;
```

Comment:
I feel like we're going to need slightly different logic here if port is zero and not.

- non-zero: this seems incorrect, as if the ports are different, the v4 port will be overridden here, right?
- zero: there is a small chance the ephemeral port bound in v6 here has already been bound by some other process on the v4 stack, and so the v4 bind will fail. Should we have a small retry loop here to get around this?
- mixed: what do we do if one of the stacks is zero, but the other is specified?

Reply:
Maybe we don't have a use-case for the ports to be different, in which case maybe we should assert they are the same. Then we just have the question of the retry loop for the ephemeral version.

Reply:
The ports should not be different. I've added a check for this and updated the documentation for the function.

You are right that we should retry if the same port isn't available for IPv4 and IPv6. I've added a retry loop that will give us a few attempts at picking a port. That's better than the user having to restart Holochain they find a port that works!